ARCs/arc-0063.md

```diff
@@ -0,0 +1,322 @@
+---
+arc: 63
+title: Lsig Plug-In Signer for Msig Vault
+description: Delegated multisig-account controlled by one account
+author: Stéphane BARROSO (@SudoWeezy)
+discussions-to: https://github.com/algorandfoundation/ARCs/issues/303
+status: Draft
+type: Standards Track
+category: ARC
+created: 2024-07-16
+---
+
+## Abstract
+
+This ARC proposes a method for creating a delegated multisig account controlled by one account and a Logic Signature (Lsig).
+
+## Motivation
+
+The motivation behind this ARC is to extend Algorand account features by enabling third-party "Plug-Ins" using a combination of delegated Lsig and Multi-Signature accounts, which act as vaults. This approach allows anyone to sign the Lsig for the vault, while maintaining security and control through a classic algorand account.
+
+## Specification
+
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href = "https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>
+
+### Components
+
+1. **Owner Account**: Vault's Owner
+1. **Lsig Plug-In**: Provided by a third party.
+1. **Plug-In Signer**: Created by generating a new key pair
+1. **1/2 Msig Account**: Comprises the owner's address, and the plug-in signer.
+
+### Implementation 1
+
+We will use the following Logic plug-in for illustrative purposes:
+
+**DO NOT USE IN PRODUCTION**
+
+```python
+teal_program = """
+#pragma version 10
+txn TypeEnum
+pushint 4
+==
+txn AssetAmount
+pushint 0
+==
+&&
+txn RekeyTo
+global ZeroAddress
+==
+&&
+txn Fee
+global MinTxnFee
+==
+&&
+return
+"""
+compiled_program = client.compile(teal_program)
+program = base64.b64decode(compiled_program["result"])
+lsig = transaction.LogicSigAccount(program)
+```
+
+> This give opt-in control over the signer
+
+1. **Generate Plug-In Signer**:
+    - Generate a random new account that we will only use once.
+
+    ```python
+    plug_in_sk, plug_in_addr  = account.generate_account()
+    ```
+
+2. **Sign Lsig with Plug-In Signer**:
+    - Sign the Lsig using the plug-in signer.
+    - Publish the public signature on the indexer.
+
+    ```python
+        public_key, secret_key = nacl.bindings.crypto_sign_seed_keypair(base64.b64decode(plug_in_sk)[: constants.key_len_bytes])
+        message = constants.logic_prefix + program
+        raw_signed = nacl.bindings.crypto_sign(message, secret_key)
+        crypto_sign_BYTES = nacl.bindings.crypto_sign_BYTES
+        signature = nacl.encoding.RawEncoder.encode(raw_signed[:crypto_sign_BYTES])
+        plug_in_public_sig = base64.b64encode(signature).decode()
```

Comment:
From my understanding, this `(LogicSigAccount(program).sign(secret_key))` method does not return the public signature `def sign(self, private_key: str) -> None:`, it will instead store it inside an lsig object like this:
`lsig = transaction.LogicSigAccount(base64.b64decode(compiled_program["result"]))`
Then, after we can sign it `lsig.sign(secret_key)`
But to extract the public signature, we shoud do the following:
`plug_in_public_sig = lsig.lsig.sig`
It works but it seems a bit more "magical"